converter/internal/test_common/testing.go

```diff
@@ -0,0 +1,138 @@
+
+				// Skip Info level diags for this testing. These would create
+				// a lot of unnecessary noise.
+				actualDiags.RemoveDiagsBySeverity(diag.SeverityLevelInfo)
+
+				expectedDiags := getExpectedDiags(t, strings.TrimSuffix(path, sourceSuffix)+diagsSuffix)
+				validateDiags(t, expectedDiags, actualDiags)
+
+				expectedRiver := getExpectedRiver(t, path, sourceSuffix)
+				validateRiver(t, expectedRiver, actualRiver)
+			})
+		}
+
+		return nil
+	}))
+}
+
+// getSourceContents reads the source file and retrieve its contents.
+func getSourceContents(t *testing.T, path string) []byte {
+	sourceBytes, err := os.ReadFile(path)
+	require.NoError(t, err)
+	return sourceBytes
+}
+
+// getTestCaseName gets the test case name based on the path and source suffix.
+func getTestCaseName(path string, sourceSuffix string) string {
+	caseName := filepath.Base(path)
+	return strings.TrimSuffix(caseName, sourceSuffix)
+}
+
+// getExpectedDiags will retrieve any expected diags for the test.
+func getExpectedDiags(t *testing.T, diagsFile string) []string {
+	expectedDiags := []string{}
+	if _, err := os.Stat(diagsFile); err == nil {
+		errorBytes, err := os.ReadFile(diagsFile)
+		require.NoError(t, err)
+		errorsString := string(normalizeLineEndings(errorBytes))
+		expectedDiags = strings.Split(errorsString, "\n")
+
+		// Some error messages have \n in them and need this
+		for ix := range expectedDiags {
+			expectedDiags[ix] = strings.ReplaceAll(expectedDiags[ix], "\\n", "\n")
+		}
+	}
+
+	return expectedDiags
+}
+
+// validateDiags makes sure the expected diags and actual diags are a match
+func validateDiags(t *testing.T, expectedDiags []string, actualDiags diag.Diagnostics) {
+	for ix, diag := range actualDiags {
+		if len(expectedDiags) > ix {
+			require.Equal(t, expectedDiags[ix], diag.String())
+		} else {
+			require.Fail(t, "unexpected diag count reach for diag: "+diag.String())
+		}
+	}
+
+	// If we expect more diags than we got
+	if len(expectedDiags) > len(actualDiags) {
+		require.Fail(t, "missing expected diag: "+expectedDiags[len(actualDiags)])
+	}
+}
+
+// normalizeLineEndings will replace '\r\n' with '\n'.
+func normalizeLineEndings(data []byte) []byte {
+	normalized := bytes.ReplaceAll(data, []byte{'\r', '\n'}, []byte{'\n'})
+	return normalized
+}
+
+// getExpectedRiver reads the expected river output file and retrieve its contents.
+func getExpectedRiver(t *testing.T, path string, sourceSuffix string) []byte {
+	outputFile := strings.TrimSuffix(path, sourceSuffix) + flowSuffix
+	if _, err := os.Stat(outputFile); err == nil {
+		outputBytes, err := os.ReadFile(outputFile)
+		require.NoError(t, err)
+		return normalizeLineEndings(outputBytes)
+	}
+
+	return nil
+}
+
+// validateRiver makes sure the expected river and actual river are a match
+func validateRiver(t *testing.T, expectedRiver []byte, actualRiver []byte) {
+	if len(expectedRiver) > 0 {
+		if !reflect.DeepEqual(expectedRiver, actualRiver) {
```

Comment:
@thampiotr I added this extra condition before printing because 1 failure was printing out all expected river output and it was confusing to parse which went wrong.

Reply:
I also do like having this print in this case a LOT 🎉 . I'm considering if we should do similar smarter printing for the diags testing as well, but that can come later too if we decide we want it.